network/src/commonMain/kotlin/dev/icerock/moko/network/features/HttpRetrySendFeature.kt

```diff
@@ -0,0 +1,80 @@
+ */
+
+package dev.icerock.moko.network.features
+
+import io.ktor.client.HttpClient
+import io.ktor.client.features.HttpClientFeature
+import io.ktor.client.request.HttpRequestBuilder
+import io.ktor.client.request.HttpSendPipeline
+import io.ktor.client.statement.HttpStatement
+import io.ktor.util.AttributeKey
+import io.ktor.utils.io.errors.IOException
+import kotlinx.coroutines.delay
+
+/**
+@param delayGetter the function for calculation a delay between failed request and next request
+@param maxAmountRetrying the max amounts retrying requests
+@param isShouldRetryRequest the function for condition a retrying request. By default it check `e is IOException `
+ */
+class HttpRetrySendFeature(
+    private val delayGetter: (lastDelayInMillisecond: Long, timeRetrying: Int) -> Long,
+    private val maxAmountsRetrying: Int,
+    private val isShouldRetryRequest: (e: Throwable) -> Boolean
+) {
+
+    class Config {
+        var onGetDelay: (lastDelayInMillisecond: Long, timeRetrying: Int) -> Long =
+            { _, _ -> 2_000L }
+        var maxAmountRetrying: Int = 3 // first request + three retrying requests,
+        var onConditionRetrying: (e: Throwable) -> Boolean = { e -> e is IOException }
+        fun build() = HttpRetrySendFeature(onGetDelay, maxAmountRetrying, onConditionRetrying)
+    }
+
+    companion object Feature : HttpClientFeature<Config, HttpRetrySendFeature> {
+        private const val LAST_DELAY_HEADER = "HttpRetrySendFeature-Last-Delay"
+        private const val RETRY_COUNTER_HEADER = "HttpRetrySendFeature-Retry-Counter"
+
+        override val key: AttributeKey<HttpRetrySendFeature> = AttributeKey("HttpRetrySendFeature")
+
+        override fun install(feature: HttpRetrySendFeature, scope: HttpClient) {
+            scope.sendPipeline.intercept(HttpSendPipeline.Before) {
+                val counter = context.headers[RETRY_COUNTER_HEADER]?.toInt() ?: 0
+                val lastDelay = context.headers[LAST_DELAY_HEADER]?.toLong()
+                    ?: feature.delayGetter(0L, counter)
+                if (counter < feature.maxAmountsRetrying) {
+                    try {
+                        context.headers.remove(LAST_DELAY_HEADER)
+                        context.headers.remove(RETRY_COUNTER_HEADER)
+                        proceed()
+                    } catch (e: Throwable) {
+                        if (feature.isShouldRetryRequest(e)) {
+                            val requestBuilder = HttpRequestBuilder().takeFrom(context)
+                            val indexRetrying = counter + 1
+                            val nextDelay = feature.delayGetter(lastDelay, counter)
+                            requestBuilder.headers[RETRY_COUNTER_HEADER] = indexRetrying.toString()
+                            requestBuilder.headers[LAST_DELAY_HEADER] = nextDelay.toString()
+                            delay(nextDelay)
+                            finish()
```

Comment:
for what we finish pipeline? later called proceed

Reply:
When finish a pipeline without a calling of proceedWith the ktor throw  error:
` java.lang.IllegalStateException: Failed to execute send pipeline. Expected to got [HttpClientCall], but received EmptyContent
        at io.ktor.client.features.HttpSend$DefaultSender.execute(HttpSend.kt:120)
        at io.ktor.client.features.HttpSend$DefaultSender$execute$1.invokeSuspend(Unknown Source:12)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:238)
        at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:194)
        at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:67)
        at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:144)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:238)
        at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:194)
        at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:67)
        at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:144)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:238)
        at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:194)
        at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:67)
        at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:144)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:238)
        at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:194)
        at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:67)
        at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:144)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:238)
        at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:194)
        at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:67)
        at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:144)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:238)
        at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:194)
        at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:67)
        at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:144)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7403)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:935)`